scvi/external/cell2location/_module.py

```diff
@@ -0,0 +1,336 @@
+
+    if shape is None:
+        alpha = torch.tensor(alpha)
+        beta = torch.tensor(beta)
+    else:
+        alpha = torch.ones(shape) * torch.tensor(alpha)
+        beta = torch.ones(shape) * torch.tensor(beta)
+    return dist.Gamma(alpha, beta)
+
+class LocationModelLinearDependentWMultiExperiment(PyroBaseModuleClass):
+    
+    def __init__(self, n_obs, n_var, n_fact, n_exper, batch_size,
+                 cell_state_mat, n_comb: int = 50,
+                 m_g_gene_level_prior={'mean': 1 / 2, 'sd': 1 / 4},
+                 m_g_gene_level_var_prior={'mean_var_ratio': 1},
+                 cell_number_prior={'N_cells_per_location': 8,
+                               'A_factors_per_location': 7,
+                               'Y_combs_per_location': 2.5},
+                 cell_number_var_prior={'N_cells_mean_var_ratio': 1,
+                                   'A_factors_mean_var_ratio': 1,
+                                   'Y_combs_mean_var_ratio': 1},
+                 alpha_g_phi_hyp_prior={'mean': 3, 'sd': 1},
+                 gene_add_alpha_hyp_prior={'mean': 3, 'sd': 1},
+                 gene_add_mean_hyp_prior={'alpha': 1, 'beta': 100},
+                 w_sf_mean_var_ratio=5):
+
+        super().__init__()
+        
+        self.n_obs = n_obs
+        self.n_var = n_var
+        self.n_fact = n_fact
+        self.n_exper = n_exper
+        self.batch_size = batch_size
+        self.n_comb = n_comb
+        
+        for k in m_g_gene_level_var_prior.keys():
+            m_g_gene_level_prior[k] = m_g_gene_level_var_prior[k]
+        for k in m_g_gene_level_prior.keys():
+            m_g_gene_level_prior[k] = np.array(m_g_gene_level_prior[k]).reshape((1, 1))
+
+        self.m_g_gene_level_prior = m_g_gene_level_prior
+        self.alpha_g_phi_hyp_prior = alpha_g_phi_hyp_prior
+        self.w_sf_mean_var_ratio = w_sf_mean_var_ratio
+        self.gene_add_alpha_hyp_prior = gene_add_alpha_hyp_prior
+        self.gene_add_mean_hyp_prior = gene_add_mean_hyp_prior
+        
+        cell_number_prior['factors_per_combs'] = (cell_number_prior['A_factors_per_location'] /
+                                                  cell_number_prior['Y_combs_per_location'])
+        for k in cell_number_var_prior.keys():
+            cell_number_prior[k] = cell_number_var_prior[k]
+        for k in cell_number_prior.keys():
+            cell_number_prior[k] = np.array(cell_number_prior[k]).reshape((1, 1))
+        self.cell_number_prior = cell_number_prior
+        
+        self.cell_state_mat = cell_state_mat
+        self.register_buffer("cell_state", torch.tensor(cell_state_mat.T))
+        
+        self.guide = AutoNormal(self.model, init_loc_fn=init_to_mean, 
+                                create_plates=self.create_plates)
+        # replace exp transform with softplus https://github.com/pyro-ppl/numpyro/issues/855
+        #self.guide = ExpToSoftplusHandler(self.guide)
+
```

Comment:
couple of comments here:

1. All prior params need to be device agnostic (see `register_buffer` in pytorch) there are also other ways to do this which I'll describe later
2. You should see our bayesian regression example in our tests. We split the logic of the model and the guide, and wrap into the PyroBaseModuleClass. Your model should be a `PyroModel` class like in that example.

Reply:
So you would do 2 classes/functions (the question is about where to add `create_plates`)?
1. `class LocationModelLinearDependentWMultiExperimentModel(PyroModel)` which implements model function as `forward` method
2. `class LocationModelLinearDependentWMultiExperiment(PyroBaseModuleClass)` which initialised model and guide:
```python
class LocationModelLinearDependentWMultiExperiment(PyroBaseModuleClass):

    def init(self)
        self.model = LocationModelLinearDependentWMultiExperimentModel(..., create_plates=self.create_plates)
        self.guide = AutoNormal(..., create_plates=self.create_plates)

    def create_plates(self, ...):
        ...
```